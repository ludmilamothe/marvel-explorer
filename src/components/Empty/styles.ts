import styled, { css } from 'styled-components';
import media from 'styled-media-query';

export const Container = styled.div`
  width: 100%;
  max-width: 44rem;
  display: flex;
  flex-direction: column;
  align-items: center;
`;

export const Image = styled.img`
  width: 100%;
`;

export const Content = styled.div`
  ${({ theme }) => css`
    color: ${theme.colors.white};
    font-size: ${theme.font.sizes.medium};
    max-width: 40rem;
    text-align: center;

    ${media.greaterThan('small')`
      font-size: ${theme.font.sizes.large};
    `}
  `}
`;
