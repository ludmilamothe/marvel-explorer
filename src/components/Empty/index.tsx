import * as S from './styles';

export type EmptyProps = {
  imagePath?: string;
  imageAlt?: string;
  children?: React.ReactNode;
};

const Empty = ({ imagePath, imageAlt, children }: EmptyProps) => {
  return (
    <S.Container>
      {imagePath && <S.Image src={imagePath} alt={imageAlt} />}
      {children && <S.Content>{children}</S.Content>}
    </S.Container>
  );
};

export default Empty;
