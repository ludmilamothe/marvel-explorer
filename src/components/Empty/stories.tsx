import { Meta, Story } from '@storybook/react/types-6-0';

import Empty, { EmptyProps } from '.';

export default {
  title: 'Empty',
  component: Empty,
  args: {
    imagePath: '/images/no-search.svg',
    imageAlt:
      'A magnifying glass with sad face and with some bubbles in the background',
    children: 'Some text',
  },
} as Meta<EmptyProps>;

export const Basic: Story<EmptyProps> = args => <Empty {...args} />;
