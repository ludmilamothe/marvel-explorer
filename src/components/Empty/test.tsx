import { screen } from '@testing-library/react';
import { renderWithProviders } from 'utils/test/helpers';

import Empty from '.';

const datas = {
  imagePath: '/images/no-search.svg',
  imageAlt:
    'A magnifying glass with sad face and with some bubbles in the background',
  children: 'Some text',
};

describe('<Empty />', () => {
  it('should render the empty correct', () => {
    renderWithProviders(<Empty {...datas} />);

    const image = screen.getByRole('img', { name: datas.imageAlt });
    const children = screen.getByText(datas.children);

    expect(children).toBeInTheDocument();

    expect(image).toBeInTheDocument();
    expect(image).toHaveAttribute('src', datas.imagePath);
  });

  it('should render the empty styles correct', () => {
    const { container } = renderWithProviders(<Empty {...datas} />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
