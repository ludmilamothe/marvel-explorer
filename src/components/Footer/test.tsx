import { screen } from '@testing-library/react';
import { renderWithTheme } from 'utils/test/helpers';

import Footer from '.';

describe('<Footer />', () => {
  it('should render the footer content correct', () => {
    renderWithTheme(<Footer />);

    const linkText = 'Mauro de Souza';
    const copy = 'Data provided by Marvel. Â© 2020 MARVEL';

    const linkElement = screen.getByRole('link', { name: linkText });

    expect(linkElement).toHaveAttribute(
      'href',
      'https://www.linkedin.com/in/maurodesouzaa/'
    );

    expect(linkElement).toHaveAttribute('target', '_blank');

    expect(screen.getByText(copy)).toBeInTheDocument();
  });

  it('should render the footer styles correct', () => {
    const { container } = renderWithTheme(<Footer />);

    const footer = container.firstChild;

    expect(footer).toMatchSnapshot();
  });
});
