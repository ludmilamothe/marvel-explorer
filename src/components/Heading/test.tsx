import { screen } from '@testing-library/react';
import { renderWithTheme } from 'utils/test/helpers';

import theme from 'styles/theme/default';
import Heading from '.';

describe('<Heading />', () => {
  it('should render the heading with children', () => {
    const text = 'Just a text to test';

    renderWithTheme(<Heading>{text}</Heading>);

    const heading = screen.getByRole('heading', { name: text });

    expect(heading).toBeInTheDocument();

    expect(heading).toMatchInlineSnapshot(`
      .c0 {
        color: #FFF;
        font-weight: 900;
        font-size: 2.4rem;
        max-width: 55rem;
        text-align: center;
      }

      @media (min-width:768px) {
        .c0 {
          font-size: 4.0rem;
        }
      }

      <h1
        class="c0"
      >
        Just a text to test
      </h1>
    `);
  });

  it('should render the heading with correct styles', () => {
    const text = 'Just a text to test';

    renderWithTheme(<Heading>{text}</Heading>);

    const heading = screen.getByRole('heading', { name: text });

    expect(heading).toHaveStyle({
      fontSize: theme.font.sizes.large,
      fontWeight: theme.font.xbold,
      color: theme.colors.white,
    });

    expect(heading).toHaveStyleRule('font-size', theme.font.sizes.xlarge, {
      media: '(min-width: 768px)',
    });
  });
});
