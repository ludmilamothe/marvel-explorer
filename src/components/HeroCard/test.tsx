/* eslint-disable @typescript-eslint/no-var-requires */

import { fireEvent, screen } from '@testing-library/react';
import { renderWithProviders } from 'utils/test/helpers';

import { storageKey } from 'hooks/useFavoriteHeroes';
import HeroCard from '.';

// eslint-disable-next-line @typescript-eslint/no-var-requires
const useRouter = jest.spyOn(require('next/router'), 'useRouter');
const push = jest.fn();

useRouter.mockImplementation(() => ({
  push,
}));

const heroData = {
  heroId: 123,
  heroName: 'Iron man',
  heroThumbnail:
    'http://i.annihil.us/u/prod/marvel/i/mg/9/c0/527bb7b37ff55.jpg',
};

describe('<HeroCard />', () => {
  beforeEach(() => {
    window.localStorage.clear();
  });

  it('should render the hero card correct', () => {
    renderWithProviders(<HeroCard {...heroData} />);

    expect(
      screen.getByRole('img', { name: heroData.heroName })
    ).toBeInTheDocument();

    expect(
      screen.getByRole('heading', { name: heroData.heroName })
    ).toBeInTheDocument();

    expect(screen.getByLabelText(/add hero to favorites/i)).toBeInTheDocument();
  });

  it('should render the hero card styles correct', () => {
    const { container } = renderWithProviders(<HeroCard {...heroData} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should add hero to favorites', () => {
    renderWithProviders(<HeroCard {...heroData} />);

    const addToFavoriteElement = screen.getByLabelText(
      /add hero to favorites/i
    );

    expect(addToFavoriteElement).toBeInTheDocument();
    fireEvent.click(addToFavoriteElement);

    const removeFromFavoriteElement = screen.getByLabelText(
      /remove hero to favorites/i
    );

    expect(removeFromFavoriteElement).toBeInTheDocument();
    expect(addToFavoriteElement).not.toBeInTheDocument();

    const storagedHero = JSON.parse(window.localStorage.getItem(storageKey)!);

    expect(storagedHero).toEqual([heroData]);
  });

  it('should remove hero to favorites', () => {
    window.localStorage.setItem(storageKey, JSON.stringify([heroData]));

    renderWithProviders(<HeroCard {...heroData} />);

    const removeFromFavoriteElement = screen.getByLabelText(
      /remove hero to favorites/i
    );

    expect(removeFromFavoriteElement).toBeInTheDocument();
    fireEvent.click(removeFromFavoriteElement);

    const addToFavoriteElement = screen.getByLabelText(
      /add hero to favorites/i
    );

    expect(removeFromFavoriteElement).not.toBeInTheDocument();
    expect(addToFavoriteElement).toBeInTheDocument();

    const storagedHero = JSON.parse(window.localStorage.getItem(storageKey)!);
    expect(storagedHero).toEqual([]);
  });

  it('should navigate to hero page', () => {
    const { container } = renderWithProviders(<HeroCard {...heroData} />);

    const heroCard = container.firstChild!;
    fireEvent.click(heroCard);

    expect(push).toBeCalledTimes(1);
    expect(push).toBeCalledWith(`/characters/${heroData.heroId}`);
  });
});
