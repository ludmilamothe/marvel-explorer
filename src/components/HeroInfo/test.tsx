import { fireEvent, screen } from '@testing-library/react';
import { renderWithProviders } from 'utils/test/helpers';

import HeroInfo from '.';

const props = {
  id: 123,
  name: 'iron man',
  description:
    'Inventor Tony Stark applies his genius for high-tech solutions to problems as Iron Man, the armored Avenger.',
  wikiLink: 'https://google.com',
  comics: 123,
  series: 1234,
  stories: 12345,
};

describe('<HeroInfo />', () => {
  it('should render the HeroInfo correct', () => {
    renderWithProviders(<HeroInfo {...props} />);

    const link = screen.getByRole('link', { name: /see more/i });
    expect(screen.getByText(props.name)).toBeInTheDocument();

    expect(screen.getByText(props.description)).toBeInTheDocument();
    expect(screen.getByText(props.comics)).toBeInTheDocument();

    expect(screen.getByText(props.series)).toBeInTheDocument();
    expect(screen.getByText(props.stories)).toBeInTheDocument();

    expect(link).toBeInTheDocument();
    expect(link).toHaveAttribute('href', props.wikiLink);
  });

  it('should alter icon between add/remove favorite', () => {
    renderWithProviders(<HeroInfo {...props} />);

    let addToFavoriteElement = screen.getByLabelText(/add hero to favorites/i);

    expect(addToFavoriteElement).toBeInTheDocument();
    fireEvent.click(addToFavoriteElement);

    const removeFromFavoriteElement = screen.getByLabelText(
      /remove hero to favorites/i
    );

    expect(removeFromFavoriteElement).toBeInTheDocument();
    expect(addToFavoriteElement).not.toBeInTheDocument();

    fireEvent.click(removeFromFavoriteElement);

    addToFavoriteElement = screen.getByLabelText(/add hero to favorites/i);

    expect(removeFromFavoriteElement).not.toBeInTheDocument();
    expect(addToFavoriteElement).toBeInTheDocument();
  });
});
