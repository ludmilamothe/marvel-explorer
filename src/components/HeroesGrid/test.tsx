import { renderWithProviders } from 'utils/test/helpers';

import HeroesGrid from '.';

describe('<HeroesGrid />', () => {
  it('should render the HeroesGrid styles correct', () => {
    const { container } = renderWithProviders(<HeroesGrid />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
