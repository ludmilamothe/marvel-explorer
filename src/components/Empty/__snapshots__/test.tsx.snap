// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Empty /> should render the empty styles correct 1`] = `
.c0 {
  width: 100%;
  max-width: 44rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  width: 100%;
}

.c2 {
  color: #FFF;
  font-size: 1.6rem;
  max-width: 40rem;
  text-align: center;
}

@media (min-width:768px) {
  .c2 {
    font-size: 2.4rem;
  }
}

<div
  class="c0"
>
  <img
    alt="A magnifying glass with sad face and with some bubbles in the background"
    class="c1"
    src="/images/no-search.svg"
  />
  <div
    class="c2"
  >
    Some text
  </div>
</div>
`;
